### PR TITLE
Flesh out auth example

### PR DESCRIPTION
**Pull request by chase-ok**
Create a more comprehensive authenticate and authorize example to help iterate on crate features and pain points.

### CODE REVIEW

File: examples/auth.rs [hunk 1]
```diff
@@ -1,45 +1,279 @@
...
+                            AuthenticateError::Unathenticated(message)
+                        // XX: should we add a .is_instance() alias for try_downcast().is_ok()?
+                        } else if exception.try_downcast::<Throwable, java_auth::AuthenticationExceptionInvalidSecurityToken>().execute(jvm)?.is_ok() {
+                            AuthenticateError::InvalidSecurityToken
+                        } else if exception.try_downcast::<Throwable, java_auth::AuthenticationExceptionInvalidSignature>().execute(jvm)?.is_ok() {
+                            AuthenticateError::InvalidSignature
+                        } else {
+                            let message = exception.get_message().assert_not_null().into_rust(jvm)?;
+                            AuthenticateError::InternalError(message)
+                        }
```
Comment (chase-ok) [inline examples/auth.rs:110]: Wild idea: could we add a macro'd enum of all downcastable types listed in the `java_package!` invocation? Something like

```rust
duchess::java_package! {
    package auth;
    
    // ...
    #[downcasting]
    class AuthenticationException { * }
    class Unauthenticated { * }
    class InvalidSecurityToken { * }
    class InvalidSignature { * }
}
```

which would then let this code look like

```rust
match exception.downcasting(jvm)? {
    AuthenticationExceptionDowncast::Unauthenticated(x) => { .. }
    AuthenticationExceptionDowncast::InvalidSecurityToken(_x) => { .. }
    AuthenticationExceptionDowncast::Else(x) => { .. }
}
```

**Review by nikomatsakis — COMMENTED** [review]
This is awesome! It's really helpful to have this example. This looks a bit different than I expected, so I'm going to do a bit of tinkering and post up a revised proposal.

File: examples/auth.rs (hunk 1 above)
Comment (nikomatsakis) [reply examples/auth.rs:110]: That's a neat idea. I think we could do *something* like that yes. I like the idea of trying to make it possible to make authoring the Java code a bit more "rust-y".

File: examples/auth.rs [hunk 2]
```diff
@@ -1,45 +1,279 @@
...
-    class AuthenticateResult { * }
-    class HttpAuth { * }
-    class HttpRequest { * }
+#[derive(Debug)]
+pub struct HttpRequest {
+    pub verb: String,
+    pub path: String,
+    pub body_hash: Vec<u8>,
+    pub params: HashMap<String, Vec<String>>,
+    pub headers: HashMap<String, Vec<String>>,
```
Comment (nikomatsakis) [inline examples/auth.rs:51]: I don't love that we are authoring this struct by hand. I wonder if we can add some kind of option to have duchess auto-generate it -- or just do it always for record types. Something to think about.

File: examples/auth.rs [hunk 3]
```diff
@@ -1,45 +1,279 @@
...
+        let auth = Jvm::with(|jvm| {
+            let auth = java_auth::HttpAuth::new().execute(jvm)?;
+            Ok(jvm.global(&*auth))
+        })?;
+        Ok(Self(auth))
+    }
+
+    pub fn authenticate(&self, request: &HttpRequest) -> Result<Authenticated, AuthenticateError> {
+        Jvm::with(|jvm| {
+            let request = request.into_java(jvm)?;
```
Comment (nikomatsakis) [inline examples/auth.rs:92]: It'd be cool to autogenerate this too, I think, and maybe base it on serialization? Have to think about it.

File: examples/auth.rs [hunk 4]
```diff
@@ -1,45 +1,279 @@
...
+impl HttpAuth {
+    pub fn new() -> duchess::GlobalResult<Self> {
+        let auth = Jvm::with(|jvm| {
+            let auth = java_auth::HttpAuth::new().execute(jvm)?;
+            Ok(jvm.global(&*auth))
+        })?;
+        Ok(Self(auth))
+    }
+
+    pub fn authenticate(&self, request: &HttpRequest) -> Result<Authenticated, AuthenticateError> {
```
Comment (nikomatsakis) [inline examples/auth.rs:90]: Ultimately, I would like this to be one big `execute` call.

**Review by nikomatsakis — COMMENTED** [review]
Some random notes -- mostly to myself -- I didn't have time to go into more detail.

**Pull request merged**